core/src/main/java/org/pentaho/di/core/vfs/configuration/KettleFileSystemConfigBuilderFactory.java

```diff
@@ -0,0 +1,88 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ******************************************************************************/
+
+package org.pentaho.di.core.vfs.configuration;
+
+import java.io.IOException;
+import java.lang.reflect.Method;
+
+import org.pentaho.di.core.variables.VariableSpace;
+import org.pentaho.di.core.vfs.KettleVFS;
+import org.pentaho.di.i18n.BaseMessages;
+
+/**
+ * This class supports overriding of config builders by supplying a VariableSpace containing a variable in the format of
+ * vfs.[scheme].config.parser where [scheme] is one of the VFS schemes (file, http, sftp, etc...)
+ *
+ * @author cboyden
+ */
+public class KettleFileSystemConfigBuilderFactory {
+
+  private static Class<?> PKG = KettleVFS.class; // for i18n purposes, needed by Translator2!!
+
+  /**
+   * This factory returns a FileSystemConfigBuilder. Custom FileSystemConfigBuilders can be created by implementing the
+   * {@link IKettleFileSystemConfigBuilder} or overriding the {@link KettleGenericFileSystemConfigBuilder}
+   *
+   * @see org.apache.commons.vfs.FileSystemConfigBuilder
+   *
+   * @param varSpace
+   *          A Kettle variable space for resolving VFS config parameters
+   * @param scheme
+   *          The VFS scheme (FILE, HTTP, SFTP, etc...)
+   * @return A FileSystemConfigBuilder that can translate Kettle variables into VFS config parameters
+   * @throws IOException
+   */
+  public static IKettleFileSystemConfigBuilder getConfigBuilder( VariableSpace varSpace, String scheme ) throws IOException {
+    IKettleFileSystemConfigBuilder result = null;
+
+    // Attempt to load the Config Builder from a variable: vfs.config.parser = class
+    String parserClass = varSpace.getVariable( "vfs." + scheme + ".config.parser" );
+
+    if ( parserClass != null ) {
+      try {
+        Class<?> configBuilderClass =
+          KettleFileSystemConfigBuilderFactory.class.getClassLoader().loadClass( parserClass );
+        Method mGetInstance = configBuilderClass.getMethod( "getInstance" );
+        if ( ( mGetInstance != null )
+          && ( IKettleFileSystemConfigBuilder.class.isAssignableFrom( mGetInstance.getReturnType() ) ) ) {
+          result = (IKettleFileSystemConfigBuilder) mGetInstance.invoke( null );
+        } else {
+          result = (IKettleFileSystemConfigBuilder) configBuilderClass.newInstance();
+        }
+      } catch ( Exception e ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)